File: optimization-sorting-algorithms-problems-advanced/src/sorting_algorithms.cpp
#include <iostream>
#include <vector>
#include <algorithm> // For std::sort (used for comparison)

//Helper Functions (Implementation in utils/helper.h)
// void mergeSort(std::vector<int>& arr);
// void quickSort(std::vector<int>& arr, int low, int high);
// void heapSort(std::vector<int>& arr);
// int quickSelect(std::vector<int>& arr, int left, int right, int k);


int main() {
    std::vector<int> arr = {12, 11, 13, 5, 6, 7};

    // Problem 1: Array Sorting
    std::vector<int> arr_copy1 = arr;
    std::sort(arr_copy1.begin(), arr_copy1.end()); //For Comparison. Replace with your implementation.
    std::cout << "Sorted array (std::sort): ";
    for (int x : arr_copy1) std::cout << x << " ";
    std::cout << std::endl;


    // Problem 2: Kth Smallest Element
    int k = 3;
    int kthSmallest = quickSelect(arr, 0, arr.size() - 1, k); // Replace with your implementation
    std::cout << "Kth smallest element (" << k << "): " << kthSmallest << std::endl;

    // Problem 3, 4, and 5 would go here with respective implementations.  Implementations are omitted for brevity.

    return 0;
}

File: data-structures-sorting-algorithms-enhanced/sorting_algorithms.py
```python
import random
import heapq

def merge_sort(arr):
    """Sorts an array using merge sort."""
    if len(arr) <= 1:
        return arr
    mid = len(arr) // 2
    left = merge_sort(arr[:mid])
    right = merge_sort(arr[mid:])
    return merge(left, right)

def merge(left, right):
    """Merges two sorted arrays."""
    merged = []
    i = j = 0
    while i < len(left) and j < len(right):
        if left[i] < right[j]:
            merged.append(left[i])
            i += 1
        else:
            merged.append(right[j])
            j += 1
    merged.extend(left[i:])
    merged.extend(right[j:])
    return merged

def quick_sort(arr):
    """Sorts an array using quick sort."""
    if len(arr) <= 1:
        return arr
    pivot = arr[len(arr) // 2]
    left = [x for x in arr if x < pivot]
    middle = [x for x in arr if x == pivot]
    right = [x for x in arr if x > pivot]
    return quick_sort(left) + middle + quick_sort(right)


def heap_sort(arr):
    """Sorts an array using heap sort."""
    heapq.heapify(arr)
    return [heapq.heappop(arr) for _ in range(len(arr))]

def bubble_sort(arr): #Inefficient but included for comparison.
    n = len(arr)
    for i in range(n):
        for j in range(0, n-i-1):
            if arr[j] > arr[j+1] :
                arr[j], arr[j+1] = arr[j+1], arr[j]
    return arr

def kth_smallest(arr, k):
    """Finds the kth smallest element using a sorted array."""
    return sorted(arr)[k-1]

# Add more functions for linked list sorting and string sorting here.

```